ESQL: Add Multi-Value support to GROK operator

luigidellaquila opened the pull request:
This PR adds support for multi-values to GROK command.

It works for both MV input and for MV patterns:

- **For MV input**, the result is a multi-value containing all the matches for all the inputs, eg.

```
row a = ["foo1 bar1", "foo2 bar2"] | grok a "%{WORD:x} .*"
```

returns 

```
 a                          |   x
["foo1 bar1", "foo2 bar2"]  |  ["foo1", "foo2"]
```

- **For MV patterns**, it returns the field as a multi-value, containing all the matches in the pattern eg.


```
row a = "foo bar" | grok a "%{WORD:x} %{WORD:x}"
```

returns 

```
 a        |   x
"foo bar" |  ["foo", "bar"]
```

- **For MV input and MV patterns together**, it returns the MV of all the values from all the inputs for all the matches in the pattern, eg.


```
row a = ["foo1 bar1", "foo2 bar2"] | grok a "%{WORD:x} %{WORD:x}"
```

returns 

```
 a                          |   x
["foo1 bar1", "foo2 bar2"]  |  ["foo1" "bar1", "foo2", "bar2"]
```

luigidellaquila committed:
Add MV support to GROK operator

both MV input and MV patterns

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ColumnExtractOperatorTests.java:
```diff
@@ -33,7 +33,8 @@ protected SourceOperator simpleInput(int end) {
         @Override
-        public void computeRow(BytesRef input, Block.Builder[] target) {
+        public void computeRow(BytesRefBlock inputBlock, int row, Block.Builder[] target, BytesRef spare) {
+            BytesRef input = inputBlock.getBytesRef(row, spare);
```
I've been using the word "index" or just `i` for these indices created by getting the position.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/GrokEvaluatorExtracter.java:
```diff
@@ -42,62 +45,67 @@ public GrokEvaluatorExtracter(
+            Integer blockIdx = keyToBlock.get(key);
+            positionToType[blockIdx] = type;
+
             fieldExtracters.add(config.objectExtracter(value -> {
```
Probably unrelated to this PR, but if you build the extracter via `nativeExtracter` instead of `objectExtracter` you could skip the boxing in most cases.

nik9000 replied:
And you wouldn't need `BlockUtils.appendValue(block, value, type);` and just call the append directly. It's much more code but cuts down the number of allocations and switches.

nik9000 replied:
And would avoid the `instanceOf Float` - you'd know you are in the `float` path and handle it natively.

nik9000 reviewed (approved):
Looks right. I left a comment about a more efficient way to do this. It isn't like grok is *ever* going to be that efficient, but every little bit helps.

I feel like this should have some kind of unit test somewhere. We probably should have an `OperatorTestCase` subclass for Grok.

luigidellaquila committed:
Implement review suggestions

luigidellaquila commented:
Thanks for the feedback @nik9000, I changed the evaluator to use `nativeExtracter`, hopefully it will save us some execution time.

> I feel like this should have some kind of unit test somewhere

Yes, I was thinking the same. The operator itself has a unit test already (ColumnExtractOperatorTests), but I added a specific test for GrokEvaluatorExtracter, that checks all the cases (single and multi-value for all the supported data types, non-matching patterns)

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/grok.csv-spec:
```diff
@@ -132,21 +132,21 @@ multivalueOutput
 
 # for now it calculates only based on the first value
 multivalueInput
 from employees | where emp_no <= 10006 | grok job_positions "%{WORD:a} %{WORD:b} %{WORD:c}" | sort emp_no | keep emp_no, a, b, c;
```
Please, put in `keep` `job_positions` as well so we can double check the correctness of `grok` output.

astefan reviewed (approved):
LGTM

luigidellaquila committed:
Enhance CSV tests

luigidellaquila committed:
Merge branch 'feature/esql' into esql/grok_mv

luigidellaquila committed:
Merge branch 'feature/esql' into esql/grok_mv
